sympy/functions/special/holonomic.py

```diff
@@ -0,0 +1,428 @@
+"""Holonomic Functions and Differential Operators"""
+
+from __future__ import print_function, division
+
+from sympy import symbols, Symbol, diff, S, Mul
+from sympy.polys.polytools import lcm, gcd
+from sympy.core.expr import Expr
+from sympy.printing import sstr
+from sympy.matrices import Matrix
+from sympy.core.compatibility import range
+from sympy.polys.polytools import DMP
+
+
+def DiffOperatorAlgebra(base, generator):
+    """ A function to create a Differential Operator Algebra.
+    The first arguments need to be the base polynomial ring for the algebra
+    and the second argument must be a generator.
+    Returns the class representing the operator algebra and
+    the standard operator for differentiation i.e. the `Dx` operator.
+
+    Examples
+    =======
+    from sympy.polys.domains import ZZ
+    R, Dx = DiffOperatorAlgebra
+    """
+
+    ring = DifferentialOperatorAlgebra(base, generator)
+    return (ring, ring.diff_operator)
+
+
+class DifferentialOperatorAlgebra:
+    """ The class representing Differential Operator Algebra.
+    Defined by the base polynomial ring and the generator.
+
+    The attribute diff_operator is the operator `Dx` which when acted
+    on a function does differentiation.
+    """
+
+    def __init__(self, base, generator):
+
+        self.base = base
+        self.diff_operator = DifferentialOperator([base.zero, base.one], self, generator)
+        self.gen_str = generator
+
+    def __str__(self):
+
+        string = 'Univariate Differential Operator Algebra in intermediate '\
+            + self.gen_str + ' over the base ring ' + \
+            (self.base).__str__()
+
+        return string
+
+    __repr__ = __str__
+
+
+class DifferentialOperator(Expr):
+    """
+    The operator can be created by providing a list of polynomials for each power of Dx
+    and the parent ring which must be an instance of DifferentialOperatorAlgebra.
+
+    An Operator can also be created easily using the operator `Dx`. See examples below.
+
+    Examples
+    ========
+
+    >>> from sympy.functions.special.holonomic import DifferentialOperator, DiffOperatorAlgebra
+    >>> from sympy.polys.domains import ZZ, QQ
+    >>> from sympy import symbols
+    >>> x = symbols('x')
+    >>> R, Dx = DiffOperatorAlgebra(ZZ.old_poly_ring(x),'Dx')
+
+    >>> DifferentialOperator([0,1,x**2], R)
+    (0) + (1)Dx + (x**2)Dx**2
+
+    One can use the `Dx` and represent differential operators more conveniently this way also.
+    >>> (x*Dx*x + 1 - Dx**2)**2
+    (2*x**2 + 2*x + 1) + (4*x**3 + 2*x**2 - 4)Dx + (x**4 - 6*x - 2)Dx**2 + (-2*x**2)Dx**3 + (1)Dx**4
+
+    """
+
+    _op_priority = 30
+
+    def __init__(self, list_of_poly, parent, generator=None, var=None):
+
+        if generator is None:
+            self.gen_symbol = symbols('Dx', commutative=False)
```

Comment:
It seems to me that the generator symbol would more properly belong to the parent algebra. All differential operators in the same operator algebra should share the same generator.